File: WEB-INF/classes/Repository.java

import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.sql.*;
import java.util.ArrayList;
import java.util.Arrays;


import org.json.JSONObject;

public class Repository implements IRepository {

    public static final String URL = "jdbc:mysql://localhost:3306/QuizApp";
    public static final String USER = "root";
    public static final String PASSWORD = "mySQL20@$CHANGE";
    private Connection con = null;   

    private void insertCategory(JSONObject entry) {
        try {
            // Insert Category
            String query = "INSERT INTO categories (name) VALUES (?)";
            PreparedStatement ps = con.prepareStatement(query);

            ps.setString(1, entry.getString("name"));
            ps.executeUpdate();

            if (!entry.isNull("media_id")) {
                // Associate quiz with respective media
                String mediaQuery = "INSERT INTO category_media (category_name, media_id) VALUES (?, ?)";
                PreparedStatement mediaPs = con.prepareStatement(mediaQuery);

                mediaPs.setString(1, entry.getString("name"));
                mediaPs.setBytes(2, entry.getString("media_id").getBytes());
                mediaPs.executeUpdate();
            }

        } catch (SQLException ex) {
            throw new RuntimeException("Error inserting entry into \"categories\" table", ex);
        }
    }

    private void insertQuiz(JSONObject entry) {
        try {
            // Insert Quiz
            String query = "INSERT INTO quizzes (name, category_name, description) VALUES (?, ?, ?)";
            PreparedStatement ps = con.prepareStatement(query);

            ps.setString(1, entry.getString("name"));
            ps.setString(2, entry.getString("category_name"));
            ps.setString(3, entry.getString("description"));
            ps.executeUpdate();

            if (!entry.isNull("media_id")) {
                // Associate quiz with respective media
                String mediaQuery = "INSERT INTO quiz_media (quiz_name, media_id) VALUES (?, ?)";
                PreparedStatement mediaPs = con.prepareStatement(mediaQuery);

                mediaPs.setString(1, entry.getString("name"));
                mediaPs.setBytes(2, entry.getString("media_id").getBytes());
                mediaPs.executeUpdate();
            }

        } catch (SQLException ex) {
            throw new RuntimeException("Error inserting entry into \"quizzes\" table", ex);
        }
    }

    private void insertQuestion(JSONObject entry) {
        try {
            // Insert Question
            String query = "INSERT INTO questions (id, quiz_name, question_text, question_type) VALUES (?, ?, ?, ?)";
            PreparedStatement ps = con.prepareStatement(query);

            ps.setBytes(1, entry.getString("id").getBytes());
            ps.setString(2, entry.getString("quiz_name"));
            ps.setString(3, entry.getString("question_text"));
            ps.setString(4, entry.getString("question_type"));
            ps.executeUpdate();

            if (!entry.isNull("media_id")) {
                // Associate question with respective media
                String mediaQuery = "INSERT INTO question_media (question_id, media_id) VALUES (?, ?)";
                PreparedStatement mediaPs = con.prepareStatement(mediaQuery);

                mediaPs.setBytes(1, entry.getString("id").getBytes());
                mediaPs.setBytes(2, entry.getString("media_id").getBytes());
                mediaPs.executeUpdate();
            }

        } catch (SQLException ex) {
            throw new RuntimeException("Error inserting entry into \"questions\" table", ex);
        }
    }

    private void insertAnswer(JSONObject entry) {
        try {
            // Insert Answer
            String query = "INSERT INTO answers (id, question_id, answer_text, is_correct, answer_type) VALUES (?, ?, ?, ?, ?)";
            PreparedStatement ps = con.prepareStatement(query);

            ps.setBytes(1, entry.getString("id").getBytes());
            ps.setBytes(2, entry.getString("question_id").getBytes());
            ps.setString(3, entry.getString("answer_text"));
            ps.setInt(4, entry.getInt("is_correct"));
            ps.setString(5, entry.getString("answer_type"));
            ps.executeUpdate();

            if (!entry.isNull("media_id")) {
                // Associate answer with respective media
                String mediaQuery = "INSERT INTO answer_media (answer_id, media_id) VALUES (?, ?)";
                PreparedStatement mediaPs = con.prepareStatement(mediaQuery);

                mediaPs.setBytes(1, entry.getString("id").getBytes());
                mediaPs.setBytes(2, entry.getString("media_id").getBytes());
                mediaPs.executeUpdate();
            }

        } catch (SQLException ex) {
            throw new RuntimeException("Error inserting entry into \"answers\" table", ex);
        }
    }

    private void insertMedia(JSONObject entry) {
        try {
            // Insert Media
            String query = "INSERT INTO media (id, description, media_type, media_file_path, media_filename, media_start, media_end) VALUES (?, ?, ?, ?, ?, ?, ?)";
            PreparedStatement ps = con.prepareStatement(query);

            ps.setBytes(1, entry.getString("id").getBytes());
            ps.setString(2, entry.getString("description"));
            ps.setString(3, entry.getString("media_type"));
            ps.setString(4, entry.getString("media_file_path"));
            ps.setString(5, entry.getString("media_filename"));
            ps.setInt(6, entry.getInt("media_start"));
            ps.setInt(7, entry.getInt("media_end"));
            ps.executeUpdate();

        } catch (SQLException ex) {
            throw new RuntimeException("Error inserting entry into \"media\" table", ex);
        }
    }

    private void insertWebSocketQuestion(JSONObject entry){
        try {
            // Insert WebsocketQuestion
            String query = "INSERT INTO QuestionsWithAnswers (questionId, questionText, answers, indexOfCorrect,quizName) VALUES (?, ?, ?, ?, ?)";
            PreparedStatement ps = con.prepareStatement(query);

            ps.setString(1, entry.getString("questionId"));
            ps.setString(2, entry.getString("questionText"));
            ps.setString(3, entry.getJSONArray("answers").toString());
            ps.setInt(4, entry.getInt("indexOfCorrect"));
            ps.setString(5, entry.getString("quizName"));

            ps.executeUpdate();

        } catch (SQLException ex) {
            throw new RuntimeException("Error inserting entry into \"media\" table", ex);
        }
    }

    private void updateQuiz(JSONObject updatedEntry, String pKey, String[] values) {
        try {
            StringBuilder query = new StringBuilder("UPDATE quizzes SET ");
            for (String col : values) {
                if(!col.equalsIgnoreCase("media_id")){
                    query.append(col).append("= ?,");
                }
                
            }
            int keyPos = 0;
            query.deleteCharAt(query.length() - 1);
            query.append(" WHERE name = ?");
            System.out.println(query.toString());
            PreparedStatement ps = con.prepareStatement(query.toString());
            for (int i = 1; i <= values.length; i++) {
                if(values[i - 1].equalsIgnoreCase("media_id")){
                    System.out.println("ADDING ENTRY TO QUIZ_MEDIA");
                    // Associate answer with respective media
                String mediaQuery = "INSERT INTO quiz_media (quiz_name, media_id) VALUES (?, ?)";
                PreparedStatement mediaPs = con.prepareStatement(mediaQuery);

                    mediaPs.setString(1, pKey);
                    mediaPs.setBytes(2, updatedEntry.getString("media_id").getBytes());
                    mediaPs.executeUpdate();
                } else {
                    ps.setString(i, updatedEntry.getString(values[i - 1]));
                }
                keyPos++;
                
            }
            
            ps.setString(keyPos + 1, pKey);
            ps.executeUpdate();
        } catch (SQLException ex) {
            System.out.println(ex);
            throw new RuntimeException("Error updating entry in the \"quizzes\" table", ex);
        }
    }

    private void updateQuestion(JSONObject updatedEntry, String pKey, String[] values){
        try {
            StringBuilder query = new StringBuilder("UPDATE questions SET ");
            for (String col : values) {
                if(!col.equalsIgnoreCase("media_id")){
                    query.append(col).append("= ?,");
                }
            }

            System.out.println("UPDATED ENTRY:" + updatedEntry.toString());

            int keyPos = 0;
            query.deleteCharAt(query.length() - 1);
            query.append(" WHERE id = ?");
            PreparedStatement ps = con.prepareStatement(query.toString());
            for (int i = 1; i <= values.length; i++) {
                System.out.println("VALUE" + values[i - 1]);
                if(values[i - 1].equalsIgnoreCase("media_id")){
                    System.out.println("INSERTING INTO QUESTION_MEDIA");
                    // Associate answer with respective media
                String mediaQuery = "INSERT INTO question_media (question_id, media_id) VALUES (?, ?)";
                PreparedStatement mediaPs = con.prepareStatement(mediaQuery);

                    mediaPs.setBytes(1, pKey.getBytes());
                    mediaPs.setBytes(2, updatedEntry.getString("media_id").getBytes());
                    mediaPs.executeUpdate();
                } else {
                    ps.setString(i, updatedEntry.getString(values[i - 1]));
                    keyPos++;
                }
                
            }
            ps.setBytes(keyPos + 1, pKey.getBytes());
            ps.executeUpdate();

            System.out.println("QUESTION UPDATED");
        } catch (SQLException ex) {
            throw new RuntimeException("Error updating entry in the \"question\" table", ex);
        }
    }

    private void updateMedia(JSONObject updatedEntry, String pKey, String[] values){
        try {
            StringBuilder query = new StringBuilder("UPDATE media SET ");
            for (String col : values) {
                query.append(col).append("= ?,");
            }
            query.deleteCharAt(query.length() - 1);
            query.append(" WHERE id = ?");
            PreparedStatement ps = con.prepareStatement(query.toString());
            for (int i = 1; i <= values.length; i++) {
                if(values[i - 1].equalsIgnoreCase("media_start") || values[i - 1].equalsIgnoreCase("media_end")){
                    ps.setInt(i, updatedEntry.getInt(values[i - 1]));
                } else {
                    ps.setString(i, updatedEntry.getString(values[i - 1]));
                }
                
            }
            ps.setBytes(values.length + 1, pKey.getBytes());
            ps.executeUpdate();

            System.out.println("MEDIA UPDATED");
        } catch (SQLException ex) {
            throw new RuntimeException("Error updating entry in the \"media\" table", ex);
        }
    }

    

    private String createConstructorParameters(ResultSet rs, String tableType) {
        StringBuilder parameters = new StringBuilder();
        try {
            ResultSetMetaData resultMetaData = rs.getMetaData();
            int columns = resultMetaData.getColumnCount();

            String prefix = "";
            for (int i = 1; i <= columns; i++) {
                String colName = resultMetaData.getColumnName(i);
                parameters.append(prefix).append(colName).append(":==");
                prefix = ",,,";
                if (colName.equals("id") || colName.contains("_id")) {
                    byte[] id = rs.getBytes(colName);
                    parameters.append(new String(id, StandardCharsets.UTF_8));
                } else {
                    String entry = rs.getString(colName);
                    parameters.append(entry);
                }
            }

            String mediaQuery;
            PreparedStatement psMedia = null;
            switch (tableType) {
                case "category":
                    mediaQuery = "SELECT media_id FROM category_media WHERE category_name = ?";
                    psMedia = con.prepareStatement(mediaQuery);
                    psMedia.setString(1, rs.getString("name"));
                    break;
                case "quiz":
                    mediaQuery = "SELECT media_id FROM quiz_media WHERE quiz_name = ?";
                    psMedia = con.prepareStatement(mediaQuery);
                    psMedia.setString(1, rs.getString("name"));
                    break;
                case "question":
                    mediaQuery = "SELECT media_id FROM question_media WHERE question_id = ?";
                    psMedia = con.prepareStatement(mediaQuery);
                    psMedia.setBytes(1, rs.getBytes("id"));
                    break;
                case "answer":
                    mediaQuery = "SELECT media_id FROM answer_media WHERE answer_id = ?";
                    psMedia = con.prepareStatement(mediaQuery);
                    psMedia.setBytes(1, rs.getBytes("id"));
                    break;
            }

            if (psMedia != null) {
                ResultSet rsMedia = psMedia.executeQuery();

                if (rsMedia.next()) {
                    byte[] media_id = rsMedia.getBytes("media_id");
                    parameters.append(prefix).append("media_id:==").append(new String(media_id, StandardCharsets.UTF_8));
                }
            }
        } catch (SQLException ex) {
            throw new RuntimeException("Error reading results of a select query from the database", ex);
        }

        return parameters.toString();
    }

    @Override
    public void init(String connectString) {
        try {
            Class.forName(connectString);
        } catch (ClassNotFoundException ex) {
            throw new RuntimeException("Error loading database driver", ex);
        }

        try {
            con = DriverManager.getConnection(URL, USER, PASSWORD);
        } catch (SQLException ex) {
            throw new RuntimeException("Error connecting to the database", ex);
        }
    }

    @Override
    public void close() {
        try {
            con.close();
        } catch (SQLException ex) {
            throw new RuntimeException("Error closing the connection to the database", ex);
        }
    }

    // "INSERT INTO media (id, media_type, media_file_path, media_filename,
    // media_start, media_end) VALUES (?, ?, ?, ?, ?, ?)";
    @Override
    public void insert(AClass entry) {
        String table = entry.getTableType();
        JSONObject newEntry = entry.serialize();
        switch (table) {
            case "category":
                insertCategory(newEntry);
                break;
            case "quiz":
                insertQuiz(newEntry);
                break;
            case "question":
                insertQuestion(newEntry);
                break;
            case "answer":
                insertAnswer(newEntry);
                break;
            case "media":
                insertMedia(newEntry);
                break;
            case "websocket":
                insertWebSocketQuestion(newEntry);
                break;
        }
    }

    @Override
    public void update(AClass update, String pKey, String values) {
        String table = update.getTableType();
        JSONObject updatedEntry = update.serialize();
        String[] changeColumns = values.split(",");
        switch (table) {
            case "quiz":
                updateQuiz(updatedEntry, pKey, changeColumns);
                break;
            case "media":
                updateMedia(updatedEntry, pKey, changeColumns);
                break;
            case "question":
                updateQuestion(updatedEntry, pKey, changeColumns);
                break;
            default:
                throw new RuntimeException("Error updating the database");
        }
    }

    @Override
    public void delete(String tableType, String criteria) {
        try {
            PreparedStatement deleteStatement = null;
            
            String update;
            byte[] id = null;
            switch (tableType) {
                case "category":
                    AClass cat = select("category", "name=" + criteria).get(0);
                    JSONObject catJSON = cat.serialize();
                    if(!catJSON.isNull("media_id")){
                        PreparedStatement  deleteCatMedia= con.prepareStatement("DELETE FROM category_media WHERE category_name =" + criteria);
                        deleteCatMedia.executeUpdate();
                    }
                    update = "DELETE FROM categories WHERE name=\"" + criteria + "\"";

                    break;
                case "quiz":
                System.out.println("DELETING QUIZ");
                    AClass quiz = select("quiz", "name=\"" + criteria + "\"").get(0);
                    JSONObject quizJSON = quiz.serialize();
                    System.out.println("QUIZ TO BE DELETE: " + quizJSON.toString());
                    if(!quizJSON.isNull("media_id")){
                        PreparedStatement deleteQuizMedia = con.prepareStatement("DELETE FROM quiz_media WHERE quiz_name =\"" + criteria + "\"");
                        deleteQuizMedia.executeUpdate();
                    }
                    
                    update = "DELETE FROM quizzes WHERE name=\"" + criteria + "\"";
                    break;
                case "question":
                    id = criteria.getBytes();
                    update = "DELETE FROM questions WHERE id = ? ";
                    break;
                case "answer":
                id = criteria.getBytes();
                    update = "DELETE FROM answers WHERE id= ?";
                    break;
                case "media":
                id = criteria.getBytes();
                    update = "DELETE FROM media WHERE id= ?";
                    break;
                default:
                    throw new RuntimeException("Error deleting from the database. Entered type cannot be deleted");
            }
            
            deleteStatement = con.prepareStatement(update);

            if(id != null){
                deleteStatement.setBytes(1, id);
            }

            deleteStatement.executeUpdate();
        } catch (SQLException ex) {
            ex.printStackTrace();
            throw new RuntimeException("Error deleting from the database");
        }

    }

    @Override
    public ArrayList<AClass> select(String tableType, String criteria) {
        // SELECT columns FROM table WHERE criteria
        AClassFactory factory = new AClassFactory();
        ArrayList<AClass> selectedEntries = new ArrayList<>();

        try {

            String query;
            byte[] id = null;
            String where = (criteria.equals("")) ? criteria : "WHERE " + criteria;
            switch (tableType) {
                case "category":
                    query = "SELECT * FROM categories " + where;
                    break;
                case "quiz":
                    query = "SELECT * FROM quizzes " + where;
                    break;
                case "question":
                    query = "SELECT * FROM questions " + where;
                    break;
                case "answer":
                    String [] queryCrits = criteria.split(",");
                    query = "SELECT * FROM answers WHERE question_id = ? ";
                    if(queryCrits.length > 1){
                        query = query + queryCrits[1];
                    }
                    
                    id = queryCrits[0].getBytes();
                    break;
                case "media":
                    id = criteria.getBytes();
                    query = "SELECT * FROM media WHERE id = ?";
                    break;
                default:
                    throw new RuntimeException("Error selecting from the database. Entered type cannot be select");
            }
            PreparedStatement selectStatement = con.prepareStatement(query);
            
            if (id != null) {
                selectStatement.setBytes(1, id);
            }

            ResultSet rs = selectStatement.executeQuery();
            while (rs.next()) {
                String parameters = createConstructorParameters(rs, tableType);
                AClass cat = factory.createAClass(tableType, parameters);
                selectedEntries.add(cat);
            }

            return selectedEntries;
        } catch (SQLException ex) {
            throw new RuntimeException("Error selecting from the database", ex);
        }
    }

}

File: QuizAppTest/Utilities/Repository.cs
using System.Linq;
using System.Security.Policy;
using System.Text.Json.Nodes;
using System.Text;
using System.Data;
using MySql.Data.MySqlClient;
using Org.BouncyCastle.Asn1.Ocsp;
using Microsoft.Extensions.Primitives;
using System.Runtime.InteropServices.JavaScript;

namespace QuizApp.Utilities
{
    public class Repository : IRepository
    {
        private IDbConnection? _connection;

        public void init(DatabaseUtil databaseUtil)
        {
            try
            {
                _connection = databaseUtil.GetConnection();
                _connection.Open();
            } catch(MySqlException ex)
            {
                
            }
        }


        public void close()
        {
            if (_connection != null)
            {
                _connection.Close();
            }
        }

        private void insertCategory(JsonObject entry)
        {
            if(_connection != null) 
            {
                // Insert Category
                var command = _connection.CreateCommand();
                command.CommandText = "INSERT INTO categories (name) VALUES (@name)";
                command.Parameters.Add(new MySqlParameter("@name", entry["name"]?.ToString()));
                command.ExecuteNonQuery();

                string? media_id = entry["id"]?.ToString();
                if (media_id != null && media_id != "")
                {

                    var mediaCommand = _connection.CreateCommand();
                    mediaCommand.CommandText = "INSERT INTO category_media (category_name, media_id) VALUES (@category_name, @media_id)";
                    mediaCommand.Parameters.Add(new MySqlParameter("@category_name", entry["name"]?.ToString()));
                    
                    byte[] media_idBytes = Convert.FromHexString(media_id);
                    mediaCommand.Parameters.Add(new MySqlParameter("@media_id", media_idBytes));
                    
                    mediaCommand.ExecuteNonQuery();
                }

            }
            
        }

        private void insertQuiz(JsonObject entry)
        {
            if (_connection != null)
            {
                // Insert Quiz

                var command = _connection.CreateCommand();
                command.CommandText = "INSERT INTO quizzes (name, category_name, description) VALUES (@name, @category_name, @description)";
                command.Parameters.Add(new MySqlParameter("@name", entry["name"]?.ToString()));
                command.Parameters.Add(new MySqlParameter("@category_name", entry["category_name"]?.ToString()));
                command.Parameters.Add(new MySqlParameter("@description", entry["description"]?.ToString()));
                command.ExecuteNonQuery();

                string? media_id = entry["id"]?.ToString();
                if (media_id != null && media_id != "")
                {
                    // Associate quiz with respective media
                    var mediaCommand = _connection.CreateCommand();
                    mediaCommand.CommandText = "INSERT INTO quiz_media (quiz_name, media_id) VALUES (@quiz_name, @media_id)";
                    mediaCommand.Parameters.Add(new MySqlParameter("@quiz_name", entry["name"]?.ToString()));
    
                    byte[] media_idBytes = Convert.FromHexString(media_id);
                    mediaCommand.Parameters.Add(new MySqlParameter("@media_id", media_idBytes));

                    mediaCommand.ExecuteNonQuery();
                }

            }
            
        }

        private void insertQuestion(JsonObject entry)
        {
            if (_connection != null)
            {
                // Insert Question
                var command = _connection.CreateCommand();
                command.CommandText = "INSERT INTO questions (id, quiz_name, question_text, question_type) VALUES (@id, @quiz_name, @question_text, @question_type)";

                string? id = entry["id"]?.ToString();
                byte[] idBytes = (id != null && id != "") ? Convert.FromHexString(id) : Array.Empty<byte>();
                command.Parameters.Add(new MySqlParameter("@id", idBytes));
                command.Parameters.Add(new MySqlParameter("@quiz_name", entry["quiz_name"]?.ToString()));
                command.Parameters.Add(new MySqlParameter("@question_text", entry["question_text"]?.ToString()));
                command.Parameters.Add(new MySqlParameter("@question_type", entry["question_type"]?.ToString()));
                command.ExecuteNonQuery();

                string? media_id = entry["media_id"]?.ToString();
                if (media_id != null && media_id != "")
                {
                    // Associate question with respective media
                    var mediaCommand = _connection.CreateCommand();
                    mediaCommand.CommandText = "INSERT INTO question_media (question_id, media_id) VALUES (@question_id, @media_id)";

                    mediaCommand.Parameters.Add(new MySqlParameter("@question_id", idBytes));
                    byte[] media_idBytes = Convert.FromHexString(media_id);
                    mediaCommand.Parameters.Add(new MySqlParameter("@media_id", media_idBytes));
                    mediaCommand.ExecuteNonQuery();
                }
            }

        }

        private void insertAnswer(JsonObject entry)
        {
            if (_connection != null)
            {
                // Insert Answer
                var command = _connection.CreateCommand();
                command.CommandText = "INSERT INTO answers (id, question_id, answer_text, is_correct, answer_type) VALUES (@id, @question_id, @answer_text, @is_correct, @answer_type)";
    
                string? id = entry["id"]?.ToString();
                byte[] idBytes = (id != null && id != "") ? Convert.FromHexString(id) : Array.Empty<byte>();
                command.Parameters.Add(new MySqlParameter("@id", idBytes));

                string? question_id = entry["question_id"]?.ToString();
                byte[] question_idBytes = question_id != null && question_id != "" ? Convert.FromHexString(question_id) : Array.Empty<byte>();
                command.Parameters.Add(new MySqlParameter("@question_id", question_idBytes));

                command.Parameters.Add(new MySqlParameter("@answer_text", entry["answer_text"]?.ToString()));

                string is_correct = entry["media_start"]?.ToString() ?? "0";
                command.Parameters.Add(new MySqlParameter("@is_correct", Int32.Parse(is_correct))); // Default to "0" if null
                command.Parameters.Add(new MySqlParameter("@answer_type", entry["answer_type"]?.ToString()));
                command.ExecuteNonQuery();

                string? media_id = entry["media_id"]?.ToString();
                if (media_id != null && media_id != "")
                {
                    // Associate answer with respective media
                    var mediaCommand = _connection.CreateCommand();
                    mediaCommand.CommandText = "INSERT INTO answer_media (answer_id, media_id) VALUES (@answer_id, @media_id)";
        
                    mediaCommand.Parameters.Add(new MySqlParameter("@answer_id", idBytes));
                    byte[] media_idBytes = Convert.FromBase64String(media_id);
                    mediaCommand.Parameters.Add(new MySqlParameter("@media_id", media_idBytes));
                    mediaCommand.ExecuteNonQuery();
                }
            }

        }

        private void insertMedia(JsonObject entry)
        {
            if (_connection != null)
            {
                // Insert Media
                var command = _connection.CreateCommand();
                command.CommandText = "INSERT INTO media (id, description, media_type, media_file_path, media_filename, media_start, media_end) VALUES (@id, @description, @media_type, @media_file_path, @media_filename, @media_start, @media_end)";

                string? id = entry["id"]?.ToString();
                byte[] idBytes = (id != null && id != "") ? Convert.FromHexString(id) : Array.Empty<byte>();
                command.Parameters.Add(new MySqlParameter("@id", idBytes));

                command.Parameters.Add(new MySqlParameter("@description", entry["description"]?.ToString()));
                command.Parameters.Add(new MySqlParameter("@media_type", entry["media_type"]?.ToString()));
                command.Parameters.Add(new MySqlParameter("@media_file_path", entry["media_file_path"]?.ToString()));
                command.Parameters.Add(new MySqlParameter("@media_filename", entry["media_filename"]?.ToString()));

                string media_start = entry["media_start"]?.ToString() ?? "0";
                string media_end = entry["media_end"]?.ToString() ?? "0";

                command.Parameters.Add(new MySqlParameter("@media_start", Int32.Parse(media_start))); // Default to "0" if null
                command.Parameters.Add(new MySqlParameter("@media_end", Int32.Parse(media_end))); // Default to "0" if null
                command.ExecuteNonQuery();
            }

        }

        private void insertWebSocketQuestion(JsonObject entry)
        {
            if (_connection != null)
            {
                // Insert WebsocketQuestion
                var command = _connection.CreateCommand();
                command.CommandText = "INSERT INTO QuestionsWithAnswers (questionId, questionText, answers, indexOfCorrect, quizName) VALUES (@questionId, @questionText, @answers, @indexOfCorrect, @quizName)";

                command.Parameters.Add(new MySqlParameter("@questionId", entry["questionId"]?.ToString()));
                command.Parameters.Add(new MySqlParameter("@questionText", entry["questionText"]?.ToString()));
                command.Parameters.Add(new MySqlParameter("@answers", entry["answers"]?.ToString()));

                string indexOfCorrect = entry["indexOfCorrect"]?.ToString() ?? "0";
                command.Parameters.Add(new MySqlParameter("@indexOfCorrect", Int32.Parse(indexOfCorrect))); // Default to "0" if null
                command.Parameters.Add(new MySqlParameter("@quizName", entry["quizName"]?.ToString()));
                command.ExecuteNonQuery();
            }

        }

        private string createConstructorParameters(Dictionary<String, Object> rs, string tableType)
        {
            StringBuilder parameters = new("");
                string prefix = "";
                foreach (var column in rs)
                {
                    string colName = column.Key;
                    parameters.Append(prefix).Append(colName).Append(":==");
                    prefix = ",,,";
                    if (colName == "id" || colName.Contains("_id"))
                    {
                        byte[]? id = (byte[])column.Value;
                        parameters.Append(BitConverter.ToString(id).Replace("-",""));
                    }
                    else
                    {
                        string? entry = column.Value.ToString();
                        parameters.Append(entry);
                    }
                }


                if(_connection != null)
                {
                    var command = _connection.CreateCommand();
                    switch (tableType)
                    {
                        case "category":
                            command.CommandText = "SELECT media_id FROM category_media WHERE category_name = @category_name";
                            command.Parameters.Add(new MySqlParameter("@category_name", rs["name"]));
                            break;
                        case "quiz":
                            command.CommandText = "SELECT media_id FROM quiz_media WHERE quiz_name = @quiz_name";
                            command.Parameters.Add(new MySqlParameter("@quiz_name", rs["name"]));
                            break;
                        case "question":
                            command.CommandText = "SELECT media_id FROM question_media WHERE question_id = @question_id";
                            command.Parameters.Add(new MySqlParameter("@question_id", rs["id"]));
                            break;
                        case "answer":
                            command.CommandText = "SELECT media_id FROM answer_media WHERE answer_id = @answer_id";
                            command.Parameters.Add(new MySqlParameter("@answer_id", rs["id"]));
                            break;
                        default:
                            command = null;
                            break;
                    }

                    if (command != null)
                    {
                        var mediaReader = command.ExecuteReader();

                        if (mediaReader.Read())
                        {
                            parameters.Append(prefix).Append("media_id:==").Append(BitConverter.ToString((byte[])mediaReader["media_id"]).Replace("-", ""));
                        }

                        mediaReader.Close();
                    }
                }
            

            return parameters.ToString();
        }

  
    public void insert(AClass entry)
        {
            string table = entry.getTableType();
            JsonObject newEntry = entry.serialize();
            switch (table)
            {
                case "category":
                    insertCategory(newEntry);
                    break;
                case "quiz":
                    insertQuiz(newEntry);
                    break;
                case "question":
                    insertQuestion(newEntry);
                    break;
                case "answer":
                    insertAnswer(newEntry);
                    break;
                case "media":
                    insertMedia(newEntry);
                    break;
                case "websocket":
                    insertWebSocketQuestion(newEntry);
                    break;
            }
        }


        private void updateQuiz(JsonObject updatedEntry, string pKey, string[] values)
        {
            if(_connection !=  null) 
            {
                StringBuilder query = new("UPDATE quizzes SET ");
                foreach (string col in values)
                {
                    query.Append(col).Append("= @").Append(col).Append(",");
                }
                query.Remove(query.Length - 1, 1);
                query.Append(" WHERE name = @name");
                var command = _connection.CreateCommand();
                command.CommandText = query.ToString();
                for (int i = 0; i < values.Length; i++)
                {
                    command.Parameters.Add(new MySqlParameter("@" + values[i], updatedEntry[values[i]]?.ToString()));
                }
                command.Parameters.Add(new MySqlParameter("@name", pKey));
                command.ExecuteNonQuery();
            }
            
        }

    public void update(AClass update, string pKey, string values)
        {
            string table = update.getTableType();
            JsonObject updatedEntry = update.serialize();
            string[] changeColumns = values.Split(",");
            switch (table)
            {
                case "quiz":
                    updateQuiz(updatedEntry, pKey, changeColumns);
                    break;
            }
        }

    public void delete(string tableType, string criteria)
        {
            if(_connection != null)
            {
                var command = _connection.CreateCommand();
                byte[]? id = null;
                switch (tableType)
                {
                    case "category":
                        command.CommandText = "DELETE FROM categories WHERE " + criteria;
                        break;
                    case "quiz":
                        command.CommandText = "DELETE FROM quizzes WHERE " + criteria;
                        break;
                    case "question":
                        id = Convert.FromHexString(criteria);
                        command.CommandText = "DELETE FROM questions WHERE id = @criteria ";
                        break;
                    case "answer":
                        command.CommandText = "DELETE FROM answers WHERE " + criteria;
                        break;
                    case "media":
                        command.CommandText = "DELETE FROM media WHERE " + criteria;
                        break;
                }

                if (id != null)
                {
                    command.Parameters.Add(new MySqlParameter("@criteria", id));
                }

                command.ExecuteNonQuery();
            }
            

        }


    public List<AClass> select(string tableType, string criteria)
        {
            // SELECT columns FROM table WHERE criteria
            AClassFactory factory = new();
            List<AClass> selectedEntries = [];

            if (_connection != null)
            {

                var command = _connection.CreateCommand();
                byte[]? id = null;
                string where = (criteria == "") ? criteria : "WHERE " + criteria;
                switch (tableType)
                {
                    case "category":
                        command.CommandText = "SELECT * FROM categories " + where;
                        break;
                    case "quiz":
                        command.CommandText = "SELECT * FROM quizzes " + where;
                        break;
                    case "question":
                        command.CommandText = "SELECT * FROM questions " + where;
                        break;
                    case "answer":
                        command.CommandText = "SELECT * FROM answers WHERE question_id = @criteria " + criteria.Split(",")[1];
                        id = Convert.FromHexString(criteria.Split(",")[0]);
                        break;
                    case "media":
                        id = Convert.FromHexString(criteria);
                        command.CommandText = "SELECT * FROM media WHERE id = @criteria";
                        break;
                }
               
                if (id != null)
                {
                    command.Parameters.Add(new MySqlParameter("@criteria", id));
                }

                var reader = command.ExecuteReader();
                var resultSet = ReadResults(reader);
                reader.Close();
                foreach(var result in resultSet)
                {
                    string parameters = createConstructorParameters(result, tableType);
                    AClass cat = factory.createAClass(tableType, parameters);
                    selectedEntries.Add(cat);
                }
            }

            return selectedEntries;

        }

        static List<Dictionary<string, object>> ReadResults(IDataReader reader)
        {
            var entries = new List<Dictionary<string, object>>();

            while (reader.Read())
            {
                var entry = new Dictionary<string, object>();
                for (int i = 0; i < reader.FieldCount; i++)
                {
                    entry[reader.GetName(i)] = reader.GetValue(i);
                }
                entries.Add(entry);
            }

            return entries;
        }
    }
}

    
